// @lwc/engine/lib/polyfills/event-listener/polyfill.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const window_1 = require("../../env/window");
const element_1 = require("../../env/element");
const dom_1 = require("../../env/dom");
const node_1 = require("../../env/node");
const node_2 = require("../../faux-shadow/node");
const events_1 = require("../../faux-shadow/events");
function doesEventNeedsPatch(e) {
    const originalTarget = dom_1.eventTargetGetter.call(e);
    if (originalTarget instanceof Node) {
        if ((node_1.compareDocumentPosition.call(document, originalTarget) &
            node_1.DOCUMENT_POSITION_CONTAINED_BY) !==
            0 &&
            node_2.getNodeOwnerKey(originalTarget)) {
            return true;
        }
    }
    return false;
}
function getEventListenerWrapper(fnOrObj) {
    let wrapperFn = null;
    try {
        wrapperFn = fnOrObj.$$lwcEventWrapper$$;
        if (!wrapperFn) {
            const isHandlerFunction = typeof fnOrObj === 'function';
            wrapperFn = fnOrObj.$$lwcEventWrapper$$ = function (e) {
                // we don't want to patch every event, only when the original target is coming
                // from inside a synthetic shadow
                if (doesEventNeedsPatch(e)) {
                    events_1.patchEvent(e);
                }
                return isHandlerFunction
                    ? fnOrObj.call(this, e)
                    : fnOrObj.handleEvent && fnOrObj.handleEvent(e);
            };
        }
    }
    catch (e) {
        /** ignore */
    }
    return wrapperFn;
}
function windowAddEventListener(type, fnOrObj, optionsOrCapture) {
    const handlerType = typeof fnOrObj;
    // bail if `fnOrObj` is not a function, not an object
    if (handlerType !== 'function' && handlerType !== 'object') {
        return;
    }
    // bail if `fnOrObj` is an object without a `handleEvent` method
    if (handlerType === 'object' &&
        (!fnOrObj.handleEvent || typeof fnOrObj.handleEvent !== 'function')) {
        return;
    }
    const wrapperFn = getEventListenerWrapper(fnOrObj);
    window_1.windowAddEventListener.call(this, type, wrapperFn, optionsOrCapture);
}
function windowRemoveEventListener(type, fnOrObj, optionsOrCapture) {
    const wrapperFn = getEventListenerWrapper(fnOrObj);
    window_1.windowRemoveEventListener.call(this, type, wrapperFn || fnOrObj, optionsOrCapture);
}
function addEventListener(type, fnOrObj, optionsOrCapture) {
    const handlerType = typeof fnOrObj;
    // bail if `fnOrObj` is not a function, not an object
    if (handlerType !== 'function' && handlerType !== 'object') {
        return;
    }
    // bail if `fnOrObj` is an object without a `handleEvent` method
    if (handlerType === 'object' &&
        (!fnOrObj.handleEvent || typeof fnOrObj.handleEvent !== 'function')) {
        return;
    }
    const wrapperFn = getEventListenerWrapper(fnOrObj);
    element_1.addEventListener.call(this, type, wrapperFn, optionsOrCapture);
}
function removeEventListener(type, fnOrObj, optionsOrCapture) {
    const wrapperFn = getEventListenerWrapper(fnOrObj);
    element_1.removeEventListener.call(this, type, wrapperFn || fnOrObj, optionsOrCapture);
}
addEventListener.__lwcOriginal__ = element_1.addEventListener;
removeEventListener.__lwcOriginal__ = element_1.removeEventListener;
windowAddEventListener.__lwcOriginal__ = window_1.windowAddEventListener;
windowRemoveEventListener.__lwcOriginal__ = window_1.windowRemoveEventListener;
function windowPatchListeners() {
    window.addEventListener = windowAddEventListener;
    window.removeEventListener = windowRemoveEventListener;
}
function nodePatchListeners() {
    Node.prototype.addEventListener = addEventListener;
    Node.prototype.removeEventListener = removeEventListener;
}
function apply() {
    windowPatchListeners();
    nodePatchListeners();
}
exports.default = apply;
