// @lwc/engine/lib/env/document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const language_1 = require("../shared/language");
const DocumentPrototypeActiveElement = language_1.getOwnPropertyDescriptor(Document.prototype, 'activeElement').get;
exports.DocumentPrototypeActiveElement = DocumentPrototypeActiveElement;
const elementFromPoint = language_1.hasOwnProperty.call(Document.prototype, 'elementFromPoint')
    ? Document.prototype.elementFromPoint
    : Document.prototype.msElementFromPoint; // IE11
exports.elementFromPoint = elementFromPoint;
const { createDocumentFragment, createElement, createElementNS, createTextNode, createComment, querySelector, querySelectorAll, getElementById, getElementsByClassName, getElementsByName, getElementsByTagName, getElementsByTagNameNS, } = Document.prototype;
exports.createDocumentFragment = createDocumentFragment;
exports.createElement = createElement;
exports.createElementNS = createElementNS;
exports.createTextNode = createTextNode;
exports.createComment = createComment;
exports.querySelector = querySelector;
exports.querySelectorAll = querySelectorAll;
exports.getElementById = getElementById;
exports.getElementsByClassName = getElementsByClassName;
exports.getElementsByName = getElementsByName;
exports.getElementsByTagName = getElementsByTagName;
exports.getElementsByTagNameNS = getElementsByTagNameNS;
