// @lwc/engine/lib/env/element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const language_1 = require("../shared/language");
const { hasAttribute, getAttribute, getAttributeNS, setAttribute, setAttributeNS, removeAttribute, removeAttributeNS, querySelector, querySelectorAll, getBoundingClientRect, getElementsByTagName, getElementsByClassName, getElementsByTagNameNS, } = Element.prototype;
exports.hasAttribute = hasAttribute;
exports.getAttribute = getAttribute;
exports.getAttributeNS = getAttributeNS;
exports.setAttribute = setAttribute;
exports.setAttributeNS = setAttributeNS;
exports.removeAttribute = removeAttribute;
exports.removeAttributeNS = removeAttributeNS;
exports.querySelector = querySelector;
exports.querySelectorAll = querySelectorAll;
exports.getBoundingClientRect = getBoundingClientRect;
exports.getElementsByTagName = getElementsByTagName;
exports.getElementsByClassName = getElementsByClassName;
exports.getElementsByTagNameNS = getElementsByTagNameNS;
let { addEventListener, removeEventListener } = Element.prototype;
exports.addEventListener = addEventListener;
exports.removeEventListener = removeEventListener;
/**
 * This trick to try to pick up the __lwcOriginal__ out of the intrinsic is to please
 * jsdom, who usually reuse intrinsic between different document.
 */
// @ts-ignore jsdom
exports.addEventListener = addEventListener = addEventListener.__lwcOriginal__ || addEventListener;
// @ts-ignore jsdom
exports.removeEventListener = removeEventListener = removeEventListener.__lwcOriginal__ || removeEventListener;
const innerHTMLSetter = language_1.hasOwnProperty.call(Element.prototype, 'innerHTML')
    ? language_1.getOwnPropertyDescriptor(Element.prototype, 'innerHTML').set
    : language_1.getOwnPropertyDescriptor(HTMLElement.prototype, 'innerHTML').set; // IE11
exports.innerHTMLSetter = innerHTMLSetter;
const tagNameGetter = language_1.getOwnPropertyDescriptor(Element.prototype, 'tagName').get;
exports.tagNameGetter = tagNameGetter;
const tabIndexGetter = language_1.getOwnPropertyDescriptor(HTMLElement.prototype, 'tabIndex').get;
exports.tabIndexGetter = tabIndexGetter;
const matches = language_1.hasOwnProperty.call(Element.prototype, 'matches')
    ? Element.prototype.matches
    : Element.prototype.msMatchesSelector; // IE11
exports.matches = matches;
const childrenGetter = language_1.hasOwnProperty.call(Element.prototype, 'children')
    ? language_1.getOwnPropertyDescriptor(Element.prototype, 'children').get
    : language_1.getOwnPropertyDescriptor(HTMLElement.prototype, 'children').get; // IE11
exports.childrenGetter = childrenGetter;
