// @lwc/engine/lib/env/__mocks__/element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
jest.genMockFromModule('./../element');
jest.unmock('./../element');
const getBoundingClientRect = jest.fn().mockReturnValue({
    width: 100,
    height: 100,
}); // Hacking in layout for jsdom.
exports.getBoundingClientRect = getBoundingClientRect;
var element_1 = require("./../element"); // Pass through all the functions from element, we probably don't want those mocked
exports.addEventListener = element_1.addEventListener;
exports.removeEventListener = element_1.removeEventListener;
exports.hasAttribute = element_1.hasAttribute;
exports.getAttribute = element_1.getAttribute;
exports.getAttributeNS = element_1.getAttributeNS;
exports.setAttribute = element_1.setAttribute;
exports.setAttributeNS = element_1.setAttributeNS;
exports.removeAttribute = element_1.removeAttribute;
exports.removeAttributeNS = element_1.removeAttributeNS;
exports.querySelector = element_1.querySelector;
exports.querySelectorAll = element_1.querySelectorAll;
exports.getElementsByTagName = element_1.getElementsByTagName;
exports.getElementsByClassName = element_1.getElementsByClassName;
exports.getElementsByTagNameNS = element_1.getElementsByTagNameNS;
exports.tagNameGetter = element_1.tagNameGetter;
exports.tabIndexGetter = element_1.tabIndexGetter;
exports.innerHTMLSetter = element_1.innerHTMLSetter;
exports.matches = element_1.matches;
